# Python Intermediário: Parte 1

## Visão geral sobre sequencias


Sequencias podem ser classificadas como: 

- _Sequências contêinerizadas_ (_container sequences_): guardam referências aos objetos que elas contêm, que podem ser de qualquer tipo. Exemplos são: `list`, `tuple` e `collections.deque`.

- _Sequências rasas_ (_flat sequences_): armazenam fisicamente o valor de cada item dentro de seu próprio espaço na memória, e não como objetos distintos. Exemplos são: `str`,`bytes` e `array.array`.

Outra forma de agrupar tipos de sequências é por mutabilidade: 

- _Sequências mutáveis_: podem ter seus elementos alterados. Exemplos: `list`,`bytearray`,`memoryview`.

- _Sequências imutáveis_: seus elementos são inalteráveis. Exemplos: `tuple`,`str`,`bytes`.

## Tuplas 

Tuplas são sequências imutáveis de comprimento fixo.

**Exemplo:** Crie uma tupla com 4 elementos e desempacote-a em 4 variáveis.

In [274]:
tup1 = -1,'a',12.23,'$'
a,b,c,d = tup1

tup2 = (-1,'a',12.23,'$')
e,f,g,h = tup2

print(tup1)
print(tup2) 

(-1, 'a', 12.23, '$')
(-1, 'a', 12.23, '$')


**Exemplo:** Crie uma tupla a partir de objetos existentes.

In [249]:
tuple(tup1)

(-1, 'a', 12.23, '$')

**Exemplo:** Construa tuplas aninhadas.

In [251]:
nums = (0,2,4,6,8),(1,3,5,7,9)
nums

((0, 2, 4, 6, 8), (1, 3, 5, 7, 9))

**Exemplo:** Altere elementos de uma tupla _in-place_.

In [252]:
t = tuple(['a',[1,2,3],False])

# imutabilidade!
t[2] = True

TypeError: 'tuple' object does not support item assignment

In [14]:
# alteração in-place
t[1].append(4)   
t 

('a', [1, 2, 3, 4], False)

**Exemplo:** Crie tuplas por concatenação.

In [256]:
('a',1) + ('b',2) + ('c',3)

('a', 1, 'b', 2, 'c', 3)

In [263]:
('+','-')*3

('+', '-', '+', '-', '+', '-')

**Exemplo:** Imprima tabela de valores usando desempacotamento e iteração.

In [267]:
s = [(1,2,3),(4,5,6),(7,8,9)]
s

for (m,n,p) in s:
    print('a={0}, b={1}, c={2}'.format(m,n,p))

a=1, b=2, c=3
a=4, b=5, c=6
a=7, b=8, c=9


**Exemplo:** Desempacote elementos em iteráveis de comprimento arbitrário.

In [283]:
# desempacota por star expression e ignora primeiro item
_,*restante = tup1
restante

['a', 12.23, '$']

Comentário: 

- Outros casos em que _star expressions_ são úteis envolvem dados que possuem mais de um valor atribuível a um mesmo registro (p.ex.: 1 pessoa com 2 números de telefone), ou quando se quer quebrar iteráveis em comprimentos arbitrários. 

- _Star expressions_ produzem listas.

Por exemplo:   

In [292]:
_,*m4,_ = (3,4,8,12,16,10)
m4 # múltiplos de 4

[4, 8, 12, 16]

In [302]:
_,_,_,*m5 = (4,8,12,5,10 )
m5 # múltiplos de 5


[5, 10]

In [299]:
*m6,_,_ = (6,12,5,10)
m6 # múltiplos de 6

[6, 12]

In [301]:
# 2 star expressions não são permitidas
*m6,*m5 = (6,12,5,10)

SyntaxError: multiple starred expressions in assignment (<ipython-input-301-5c1c52953256>, line 2)

>Ao usar uma _star expression_, certifique-se que o número de variáveis usadas no desempacotamento é consistente com os seus objetivos.

### Métodos de tupla

Métodos são funções atribuídas a alguns objetos. Embora tuplas sejam imutáveis, elas possuem métodos bastante úteis. Aqui, consideraremos dois:

- `count`: conta quantas vezes um dado valor aparece na tupla.
- `index`: localiza o índice de um valor dado.

**Exemplo:** Contando valores em tupla.

In [306]:
tupla = (4,3,2,1,1,2,3,4,2,2,2,3,3,3,1,3,1,2,3,4,2,1,3,4,2,2)

print(tupla.count(0))
print(tupla.count(3))
print(tupla.count(4))

0
8
4


**Exemplo:** Localizando posição.

In [307]:
tupla.index(3)

1

### _Swap_ de variáveis

O modo Pythônico de realizar uma troca de variáveis (_swap_) é o seguinte:

In [309]:
x,y = 1,2
y,x = x,y
x,y

(2, 1)

### Tuplas como registros

Tuplas servem para armazenar registros: cada item na tupla é o dado de um campo e a posição do item dá o significado. 

**Exemplo:** Tuplas como registros. 

In [319]:
# latitute e longitude da Torre Eiffel
eiffel_ll = (48.85844772530444, 2.2948031631859886)

# capitais do mundo
caps = [('Afeganistão','Cabul','Ásia'),
        ('Barbados','Bridgetown','América'),
        ('Chade','Jamena','África')]

# população indígena no Brasil por decênio
popind = [(1991,294131),(2000,734127),(2010,817963)]

for a,p in popind:
    print(f'Em {a}, o Brasil possuía {p} habitantes indígenas.')

Em 1991, o Brasil possuía 294131 habitantes indígenas.
Em 2000, o Brasil possuía 734127 habitantes indígenas.
Em 2010, o Brasil possuía 817963 habitantes indígenas.


In [321]:
print('{:14} | {:14} | {:14}'.format('PAÍS','CAPITAL','CONTINENTE'))
print('-'*(14*3+2))
fmt = '{:14} | {:14} | {:14}' 
for pais, cap, cont in caps:
    print(fmt.format(pais,cap,cont))

PAÍS           | CAPITAL        | CONTINENTE    
--------------------------------------------
Afeganistão    | Cabul          | Ásia          
Barbados       | Bridgetown     | América       
Chade          | Jamena         | África        


Comentários: 

- Neste exemplo, simulamos uma tabela usando um formato-base e criamos uma linha de cabeçalho por concatenação de string.

## Dicionários 

Um dicionário (_dict_) é uma estrutura conhecida como _tabela hash_ ou _array associativo_. É uma coleção de tamanho flexível composta de pares do tipo _chave:valor_. Criamos um `dict` por diversas formas. A mais simples é usar chaves e pares explícitos.

In [322]:
d = {} # dict vazio
d

{}

Os pares chave-valor incorporam quaisquer tipos de dados.

In [328]:
d = {'par': [0,2,4,6,8], 'ímpar': [1,3,5,7,9], 'nome':'Meu dict', 'teste': True}
d

{'par': [0, 2, 4, 6, 8],
 'ímpar': [1, 3, 5, 7, 9],
 'nome': 'Meu dict',
 'teste': True}

### Acesso a conteúdo

Para acessar o conteúdo de uma chave, indexamos pelo seu nome.

In [110]:
d['par'] 

[0, 2, 4, 6, 8]

In [329]:
d['nome']

'Meu dict'

**Exemplo:** construindo soma e multiplicação especial.

In [330]:
# dict
op = {'X' :[1,2,3], 'delta' : 0.1}

# função
def sp(op): 
    s = [x + op['delta'] for x in op['X']]
    p = [x * op['delta'] for x in op['X']]
    
    return (s,p) # retorna tupla

soma, prod = sp(op) # desempacota

for i,s in enumerate(soma):
    print(f'pos({i}) | Soma = {s} | Prod = {prod[i]}')

pos(0) | Soma = 1.1 | Prod = 0.1
pos(1) | Soma = 2.1 | Prod = 0.2
pos(2) | Soma = 3.1 | Prod = 0.30000000000000004


### Inserção de conteúdo

In [331]:
# apensa variáveis
op[1] = 3 
op['novo'] = (3,4,1) 
op

{'X': [1, 2, 3], 'delta': 0.1, 1: 3, 'novo': (3, 4, 1)}

### Alteração de conteúdo

In [332]:
op['novo'] = [2,1,4] # sobrescreve
op

{'X': [1, 2, 3], 'delta': 0.1, 1: 3, 'novo': [2, 1, 4]}

### Deleção de conteúdo com `del` e `pop`

In [333]:
del op[1] # deleta chave 
op  

{'X': [1, 2, 3], 'delta': 0.1, 'novo': [2, 1, 4]}

In [334]:
novo = op.pop('novo') # retorna e simultaneamente deleta
novo

[2, 1, 4]

In [335]:
op  

{'X': [1, 2, 3], 'delta': 0.1}

### Listagem de chaves e valores

Usamos os métodos `keys()` e `values()` para listar chaves e valores.

In [342]:
arit = {'soma': '+', 'subtr': '-', 'mult': '*', 'div': '/'} # dict

k = list(arit.keys())
print(k)
val = list(arit.values())
print(val)
for v in range(len(arit)):
    print(f'A operação \'{k[v]}\' de "arit" usa o símbolo \'{val[v]}\'.')    

['soma', 'subtr', 'mult', 'div']
['+', '-', '*', '/']
A operação 'soma' de "arit" usa o símbolo '+'.
A operação 'subtr' de "arit" usa o símbolo '-'.
A operação 'mult' de "arit" usa o símbolo '*'.
A operação 'div' de "arit" usa o símbolo '/'.


### Combinando dicionários

Usamos `update` para combinar dicionários. Este método possui um resultado similar a `extend`, usado em listas.

In [343]:
pot = {'pot': '**'}
arit.update(pot)
arit

{'soma': '+', 'subtr': '-', 'mult': '*', 'div': '/', 'pot': '**'}

### Dicionários a partir de sequencias

Podemos criar dicionários a partir de sequencias existentes usando `zip`.

In [347]:
arit = ('soma', 'subtr', 'mult', 'div', 'pot')
ops = ('+', '-', '*', '/', '**')

dict_novo = {}

for chave,valor in zip(arit,ops):
    dict_novo[chave] = valor
    
dict_novo

{'soma': '+', 'subtr': '-', 'mult': '*', 'div': '/', 'pot': '**'}

Visto que um `dict` é composto de várias tuplas de 2, podemos criar um  de maneira ainda mais simples.

In [348]:
dict_novo = dict(zip(arit,ops)) # visto que dicts
dict_novo

{'soma': '+', 'subtr': '-', 'mult': '*', 'div': '/', 'pot': '**'}

## Conjuntos

Um conjunto (_set_) é uma coleção não ordenada de elementos únicos. Eles podem ser criados através da função `set` ou de uma expressão literal com um par de chaves `{}`. Eles são parecidos com _dicts_, porém não possuem chaves (_keys_).


**Exemplo:** Crie o conjunto dos números pares positivos e menores do que 15.

In [361]:
pq1 = set(range(0,15,2))
pq2 = {0,2,4,6,8,10,12,14}

Discussão: 

- Em `pq1`, criamos o conjunto por função, usando `range` para gerar os números.
- Em `pq2`, criamos o conjunto literalmente, por extensão.

In [363]:
{1,2,2,3,3,4,4,4} # 'set' possui unicidade de elementos

{1, 2, 3, 4}

### Operações com conjuntos 

A seguir mostraremos uma série de operações com conjuntos. Considere os seguintes conjuntos.

In [364]:
A = {1,2,3}
B = {3,4,5}
C = {6}

#### União de conjuntos

In [365]:
A.union(B) # união

{1, 2, 3, 4, 5}

In [366]:
A | B # união com operador alternativo ('ou')

{1, 2, 3, 4, 5}

#### Atualização de conjuntos (união)

A união *in-place* de dois conjuntos pode ser feita com `update`.

In [367]:
C

{6}

In [371]:
C.update(B) # C é atualizado com elementos de B
C 

{3, 4, 5, 6}

In [372]:
C.union(A) # conjunto união com A

{1, 2, 3, 4, 5, 6}

In [373]:
C # os elementos de A não foram atualizados em C

{3, 4, 5, 6}

A atualização da união possui a seguinte forma alternativa com `|=`.

In [374]:
C |= A # elementos de A atualizados em C
C

{1, 2, 3, 4, 5, 6}

#### Interseção de conjuntos

In [375]:
A.intersection(B) # interseção

{3}

In [376]:
A & B # interseção com operador alternativo ('e')

{3}

#### Atualização de conjuntos (interseção)

A interseção *in-place* de dois conjuntos pode ser feita com `intersection_update`.

In [377]:
D = {1, 2, 3, 4}
E = {2, 3, 4, 5}

In [378]:
D.intersection(E) # interseção com E

{2, 3, 4}

In [379]:
D # D inalterado

{1, 2, 3, 4}

In [380]:
D.intersection_update(E) 
D # D alterado

{2, 3, 4}

A atualização da interseção possui a seguinte forma alternativa com `&=`.

In [381]:
D &= E
D

{2, 3, 4}

#### Diferença entre conjuntos

In [139]:
A

{1, 2, 3}

In [140]:
D

{2, 3, 4}

In [141]:
A.difference(D) # apenas elementos de A

{1}

In [142]:
D.difference(A) # apenas elementos de D

{4}

In [143]:
A - D # operador alternativo 

{1}

In [144]:
D - A 

{4}

#### Atualização de conjuntos (diferença)

A interseção *in-place* de dois conjuntos pode ser feita com `difference_update`.

In [382]:
D = {1, 2, 3, 4}
E = {1, 2, 3, 5}

In [146]:
D

{1, 2, 3, 4}

In [383]:
D.difference(E)
D

{1, 2, 3, 4}

In [384]:
D.difference_update(E)
D

{4}

A atualização da diferença possui a seguinte forma alternativa com `-=`.

In [385]:
D -= E
D

{4}

#### Adição ou remoção de elementos

In [387]:
A

{1, 2, 3}

In [388]:
A.add(4) # adiciona 4 a A
A

{1, 2, 3, 4}

In [389]:
B

{3, 4, 5}

In [390]:
B.remove(3) # remove 3 de B
B

{4, 5}

#### Reinicialização de um conjunto (vazio)

Podemos remover todos os elementos de um conjunto com `clear`, deixando-o em um estado vazio.

In [391]:
A

{1, 2, 3, 4}

In [392]:
A.clear()
A # A é vazio

set()

In [393]:
len(A) # 0 elementos

0

#### Continência

Podemos verificar se um conjunto $A$ é subconjunto de (está contido em) outro conjunto $B$ ($A \subseteq B$) ou se $B$ é um superconjunto para (contém) $A$ ($B \supseteq A$) com `issubset` e  `issuperset`. 

In [396]:
B

{4, 5}

In [395]:
C

{1, 2, 3, 4, 5, 6}

In [397]:
B.issubset(C) # B está contido em C

True

In [398]:
C.issuperset(B) # C contém B

True

#### Subconjuntos e subconjuntos próprios

Podemos usar operadores de comparação entre conjuntos para verificar continência.

- $A \subseteq B$: $A$ é subconjunto de $B$
- $A \subset B$: $A$ é subconjunto próprio de $B$ ($A$ possui elementos que não estão em $B$)

In [399]:
{1,2,3} <= {1,2,3} # subconjunto

True

In [400]:
{1,2} < {1,2,3} # subconjunto próprio

True

In [401]:
{1,2,3} > {1,2}

True

In [402]:
{1,2} >= {1,2,3}

False

#### Disjunção

Dois conjuntos são disjuntos se sua interseção é vazia. Podemos verificar a disjunção com `isdisjoint`

In [403]:
E

{1, 2, 3, 5}

In [404]:
G = {1,2,3,4}
G

{1, 2, 3, 4}

In [405]:
E.isdisjoint(G) # 1,2,3 são comuns

False

#### Igualdade entre conjuntos

Dois conjuntos são iguais se contém os mesmos elementos.

In [406]:
H = {3,'a', 2}
I = {'a',2, 3}
J = {1,'a'}

In [407]:
H == I

True

In [408]:
H == J

False

In [410]:
{1,2,2,3} == {3,3,3,2,1} # lembre-se da unicidade

True

## Compreensões de lista (_listcomps_)

**Exemplo:** Construa uma lista dos códigos _Unicode_ correspondentes dos caracteres de uma _string_.

In [172]:
palavra = '!@#$%'
cods = []
for c in palavra:
    cods.append(ord(c))
cods    

[33, 64, 35, 36, 37]

Discussão:

- `ord` produz o número ordinal correspondente de um caracter segundo o padrão Unicode. Para saber o código hexadecimal correspondente, pode-se fazer `hex(ord(c))`.

**Exemplo:** Construa uma _listcomp_ correspondente.

In [414]:
palavra = '!@#$%'
cods = [ord(c) for c in palavra]
cods

[33, 64, 35, 36, 37]

In [416]:
# listcomp com quebra de linha sem `\`
[ord(c) 
 for c 
 in palavra]

[33, 64, 35, 36, 37]

**Exemplo:** Construa um produto cartesiano de pares $(a,b)$, onde $a$ é um tipo de sangue de acordo com o sistema ABO e $b$ é o Fator Rh.

In [417]:
A = ['A','B','O','AB']
Rh = ['+','-']
sangue = [(a,b) for a in A for b in Rh]
sangue

[('A', '+'),
 ('A', '-'),
 ('B', '+'),
 ('B', '-'),
 ('O', '+'),
 ('O', '-'),
 ('AB', '+'),
 ('AB', '-')]

**Exemplo:** Construa um produto cartesiano de pares $(a,b)$, onde $a$ é o FatorRh e $b$ é um tipo de sangue de acordo com o sistema ABO.

In [418]:
[(b,a) for a in A for b in Rh]

[('+', 'A'),
 ('-', 'A'),
 ('+', 'B'),
 ('-', 'B'),
 ('+', 'O'),
 ('-', 'O'),
 ('+', 'AB'),
 ('-', 'AB')]

**Exemplo:** Use _listcomps_ com condicionais para criar filtros.

In [419]:
# equipes da fórmula 1
f1 = ['Red Bull Racing','Mercedes',
 'McLaren','Ferrari','AlphaTauri',
 'Aston Martin','Alpine','Alfa Romeo Racing',
 'Williams','Haas F1 Team']

# equipes que começam com A, em maiúsculas
[team.upper() for team in f1 if team[0] == 'A']

['ALPHATAURI', 'ASTON MARTIN', 'ALPINE', 'ALFA ROMEO RACING']

In [420]:
# equipes cujo nome não inicia com A e tem menos do que 10 caracteres
[team for team in f1 
 if team[0] != 'A' and len(team) < 10]

['Mercedes', 'McLaren', 'Ferrari', 'Williams']

## Compreensões de conjuntos

Semelhantemente a _listcomps_, podemos realizar "_setcomps_".

In [424]:
# comprimentos de nomes de equipes (únicos)
{len(team) for team in f1}

{6, 7, 8, 10, 12, 15, 17}

In [430]:
[len(team) for team in f1]

[15, 8, 7, 7, 10, 12, 6, 17, 8, 12]

In [432]:
# comprimento de nomes de equipes que terminam com "ing"
{len(team) for team in f1 if team.endswith('ing')}

{15, 17}

## Compreensões de dicionários

Compreensões de dicionários, ou "_dictcomps_", possuem resultados similares às anteriores e são convenientes para criar _dicts_.

In [434]:
{k:v for k,v in enumerate(f1)}

{0: 'Red Bull Racing',
 1: 'Mercedes',
 2: 'McLaren',
 3: 'Ferrari',
 4: 'AlphaTauri',
 5: 'Aston Martin',
 6: 'Alpine',
 7: 'Alfa Romeo Racing',
 8: 'Williams',
 9: 'Haas F1 Team'}

In [435]:
# enumera a partir do índice 7, mas percorre toda a lista
{k:v for k,v in enumerate(f1,7)}

{7: 'Red Bull Racing',
 8: 'Mercedes',
 9: 'McLaren',
 10: 'Ferrari',
 11: 'AlphaTauri',
 12: 'Aston Martin',
 13: 'Alpine',
 14: 'Alfa Romeo Racing',
 15: 'Williams',
 16: 'Haas F1 Team'}

## _Sorted_ 

- Listas em Python possuem um método chamado `sort` com o qual podemos realizar uma ordenação _in-place_ de uma lista. 

- A função `sorted` permite que façamos o mesmo criando uma nova lista e, dessa maneira, não alterando a lista original.

In [444]:
# copia lista f1
f1c = f1.copy()
f1c

['Red Bull Racing',
 'Mercedes',
 'McLaren',
 'Ferrari',
 'AlphaTauri',
 'Aston Martin',
 'Alpine',
 'Alfa Romeo Racing',
 'Williams',
 'Haas F1 Team']

In [445]:
# lista ordenada
f1c.sort()
f1c

['Alfa Romeo Racing',
 'AlphaTauri',
 'Alpine',
 'Aston Martin',
 'Ferrari',
 'Haas F1 Team',
 'McLaren',
 'Mercedes',
 'Red Bull Racing',
 'Williams']

In [446]:
# lista f1 ordenada
f1c2 = sorted(f1)
f1c2

['Alfa Romeo Racing',
 'AlphaTauri',
 'Alpine',
 'Aston Martin',
 'Ferrari',
 'Haas F1 Team',
 'McLaren',
 'Mercedes',
 'Red Bull Racing',
 'Williams']

In [447]:
# lista original intacta
f1

['Red Bull Racing',
 'Mercedes',
 'McLaren',
 'Ferrari',
 'AlphaTauri',
 'Aston Martin',
 'Alpine',
 'Alfa Romeo Racing',
 'Williams',
 'Haas F1 Team']

In [448]:
# ordenação re
sorted(f1,reverse=True)

['Williams',
 'Red Bull Racing',
 'Mercedes',
 'McLaren',
 'Haas F1 Team',
 'Ferrari',
 'Aston Martin',
 'Alpine',
 'AlphaTauri',
 'Alfa Romeo Racing']

## Aleatoriedade

- Números aleatórios, ou também chamados de randômicos, são úteis para fabricar experimentos estatísticos, realizar amostragem e desenvolver uma série de estudos envolvendo probabilidade.  

- Em Python, podemos usar o módulo `random` para atingir uma diversidade de propósitos em ciência de dados. Vamos explorar algumas funções para geração aleatória: `choice`, `random`, `randint`, `shuffle` e `sample`.

In [189]:
import random

**Exemplo:** Crie um experimento de lançamento de dados e realize uma escolha aleatória em $n$ lançamentos.

In [190]:
dado = list(range(1,7))

n = 4
for _ in range(n):
    print(random.choice(dado))

4
1
2
1


Comentários:

- O _loop_ será repetido $n$ vezes.
- O método `choice` escolhe um valor ao acaso entre os disponíveis na lista. 

**Exemplo:** Crie um experimento de probabilidade aleatório que determine a salinidade presente em 1 litro de água marinha (3 - 5%). 

In [191]:
na = 5
saly = []
for _ in range(na):
    saly.append(0.03 + (0.05-0.03)*random.random())
saly    

[0.0305926166842815,
 0.048945321318505396,
 0.047421318696230105,
 0.03159297358232361,
 0.03967769053476741]

Comentários: 

- Este experimento equivale a coletar 5 amostras de água marinha com salinidade variável entre 3 e 5%.
- O método `random` determina um número aleatório no intervalo [0,1).

**Exemplo:** Crie um experimento que ordena 15 arremessadores para cestas de 3 pontos durante um treinamento do time do Los Angeles Lakers (0 a 99) excluindo a lista de números reservados. Considere que se o número sorteado não estiver atribuído a nenhum jogador atual ou se já tiver sido sorteado, o técnico escolherá o próximo da lista.

Lista de números reservados: 8,13,17,19,22,24,25,32,33,34,42,44,52,99.

In [192]:
reservados = [8,13,17,19,22,24,25,32,33,34,42,44,52,99]

shooter = []
while len(shooter) != 15:
    x = random.randint(0,99)
    if x not in reservados and x not in shooter:
        shooter.append(x)    

print('Los Angeles Lakers :: 3-point shooters (Training Schedule):')
for s in shooter:
    print(s)
        

Los Angeles Lakers :: 3-point shooters (Training Schedule):
79
63
81
9
30
15
18
90
59
50
7
89
11
98
1


Comentários: 

- O _loop_ _while_ encerrará quando uma lista com 15 arremessadores estiver completa.
- A estrutura condicional previne que números já sorteados entrem na lista.
- O método `randint` escolhe um número inteiro aleatório entre 0 e 99.

**Exemplo:** Crie uma alteração no schedule de treinamento anterior a partir da última lista disponível.

In [193]:
print(random.shuffle(shooter))
shooter

None


[15, 59, 7, 63, 18, 89, 9, 30, 1, 98, 50, 79, 81, 90, 11]

Comentário: 

- O método `shuffle` reordena a lista `shooter` randomicamente e produz um valor `None` como saída. Dessa forma, geramos uma nova lista com os mesmos arremessadores pré-selecionados, mas em ordem distinta.

**Exemplo:** Construa uma tabela aleatória de 3 jogos de beisebol entre times da MLB.

In [194]:
# lê arquivo com nome de times da MLB
with open('../database/mlb-teams.csv') as f:
    mlb = f.read().splitlines()
    
for _ in range(3):
    a,b = random.sample(mlb,2)
    print(f'Game {_+1}: {a} x {b}')        

Game 1: Washington Nationals x St. Louis Cardinals
Game 2: Seattle Mariners x Houston Astros
Game 3: Atlanta Braves x St. Louis Cardinals


Comentários: 

- O método `sample` realiza uma amostragem de $k$ valores disponíveis na lista. Aqui, escolhemos $k=2$, de modo que 2 times são escolhidos ao acaso para comporem 3 jogos.

## Expressões regulares

Expressões regulares (_regular expressions_), também conhecidas como _RE_, _regexes_, ou _regex pattern_ são regras estabelecidas para localizar padrões desejados em strings. REs, na verdade, formam uma pequena linguagem de programação dentro de uma linguagem maior. 

### Padrões simples, caracterese ordinários e metacaracteres

- A tarefa mais simples que podemos executar com REs é combinar caracteres.

_Caracteres ordinários_ são basicamente todos os caracteres ASCII alfanuméricos, tais como `'b'`,  `'B'` e `'1'`.

_Caracteres especiais_ são todos os metacaracteres. A lista dos metacaracteres é a seguinte: 

```
. ^ $ * + ? { } [ ] \ | ( )
``` 

In [195]:
# imprime caracteres ASCII mais comuns 
# end = ' ' altera a terminação padrão '\n'
for i in range(33,127):
    print(chr(i), end = ' ')

! " # $ % & ' ( ) * + , - . / 0 1 2 3 4 5 6 7 8 9 : ; < = > ? @ A B C D E F G H I J K L M N O P Q R S T U V W X Y Z [ \ ] ^ _ ` a b c d e f g h i j k l m n o p q r s t u v w x y z { | } ~ 

### O módulo `re`

- Em Python, temos à disposição o módulo `re` para trabalhar com REs. 

- Vamos ler o arquivo bras-cubas.txt e salvar o texto em uma _string_.

In [196]:
with open('../database/bras-cubas.txt', 'r') as f:
    cubas = f.read()
cubas    

'Não durou muito a evocação; a realidade dominou logo; o presente expeliu o passado. Talvez eu exponha ao leitor, em algum canto deste livro, a minha teoria das edições humanas. O que por agora importa saber é que Virgília — chamava-se Virgília — entrou na alcova, firme, com a gravidade que lhe davam as roupas e os anos, e veio até o meu leito. O estranho levantou-se e saiu. Era um sujeito, que me visitava todos os dias para falar do câmbio, da colonização e da necessidade de desenvolver a viação férrea; nada mais interessante para um moribundo. Saiu; Virgília deixou-se estar de pé; durante algum tempo ficamos a olhar um para o outro, sem articular palavra. Quem diria? De dois grandes namorados, de duas paixões sem freio, nada mais havia ali, vinte anos depois; havia apenas dois corações murchos, devastados pela vida e saciados dela, não sei se em igual dose, mas enfim saciados. Virgília tinha agora a beleza da velhice, um ar austero e maternal; estava menos magra do que quando a vi, p

In [197]:
# importa o módulo
import re

As principais funções do módulo `re` podem ser resumidas ao seguinte modelo:

```python
re.f(pattern,string)
```
onde _f_ é o nome da função, _pattern_ é o padrão, isto é, a _regex_, e _string_ é a string na qual o padrão é procurado. Exploraremos algumas delas a seguir compreendendo, simultaneamente, o papel desempenhado pelos metacaracteres. 

**Exemplo:** Procure pelo padrão "que" no texto-base e conte quantas vezes ele aparece.

In [198]:
# busca todas as combinações
que = re.findall('que',cubas)
print(que)
print(len(que))

['que', 'que', 'que', 'que', 'que', 'que', 'que']
7


In [199]:
# padrão presente, mas não é palavra inteligível
re.findall('xou',cubas)

['xou']

Buscas por caracteres ordinários são sempre exatas, no sentido de que o caracter é combinado como ele é. Vejamos um exemplo que conta as vogais que aparecem no texto.

In [200]:
for vogal in ['a','e','i','o','u']:    
    print(f'A vogal \'{vogal}\' foi \
localizada {len(re.findall(vogal,cubas))} vezes no texto.')

A vogal 'a' foi localizada 143 vezes no texto.
A vogal 'e' foi localizada 104 vezes no texto.
A vogal 'i' foi localizada 65 vezes no texto.
A vogal 'o' foi localizada 83 vezes no texto.
A vogal 'u' foi localizada 43 vezes no texto.


**Exemplo:** Busque pelo padrão "Tijuca" no texto e determine suas posições de início e término.

In [201]:
# busca por combinação retornado posição
p = re.search('Tijuca',cubas)
p.start(), p.end(), p.group()

(1042, 1048, 'Tijuca')

Discussão: 

- O método `search` procura pelo padrão `'Tijuca'` em `cubas` e retorna três informações:
    - `start`: a posição inicial do padrão
    - `end`: a posição final do padrão
    - `group`: o padrão localizado
- O padrão aparece apenas uma vez no texto na fatia `cubas[1042:1048]`.

Comentários: 

- Caso houvesse repetição do padrão, apenas os índices da primeira aparição seriam retornados. Vejamos: 

In [202]:
p2 = re.search('que',cubas)
p2.start(), p2.end(), p2.group()

(179, 182, 'que')

Vimos anteriormente que há 7 aparições do padrão "que" no texto-base. Porém, apenas a primeira é retornada por `search`. Vejamos:

In [203]:
# uma maneira de mostrar que há apenas
# um 'que' até o caracter 182 de 'cubas'
# por meio de split e interseção de conjuntos
set(cubas[0:182].split(' ')).intersection({'que'})

{'que'}

Como este conjunto é unitário, há apenas um padrão 'que' até a posição 182 de `cubas`.

**Exemplo:** Busque pelo padrão "Não" no texto e verifique se ele aparece no início da string.

In [204]:
p3 = re.match('Não',cubas)
p3.start()

0

**Exemplo:** Substitua o padrão "Virgília" por "Ofélia" em toda a string do texto.

In [205]:
cubas_novo = re.sub('Virgília','Ofélia',cubas)
cubas_novo

'Não durou muito a evocação; a realidade dominou logo; o presente expeliu o passado. Talvez eu exponha ao leitor, em algum canto deste livro, a minha teoria das edições humanas. O que por agora importa saber é que Ofélia — chamava-se Ofélia — entrou na alcova, firme, com a gravidade que lhe davam as roupas e os anos, e veio até o meu leito. O estranho levantou-se e saiu. Era um sujeito, que me visitava todos os dias para falar do câmbio, da colonização e da necessidade de desenvolver a viação férrea; nada mais interessante para um moribundo. Saiu; Ofélia deixou-se estar de pé; durante algum tempo ficamos a olhar um para o outro, sem articular palavra. Quem diria? De dois grandes namorados, de duas paixões sem freio, nada mais havia ali, vinte anos depois; havia apenas dois corações murchos, devastados pela vida e saciados dela, não sei se em igual dose, mas enfim saciados. Ofélia tinha agora a beleza da velhice, um ar austero e maternal; estava menos magra do que quando a vi, pela últi

**Exemplo:** Substitua o padrão "Virgília" por "Ofélia" apenas nas duas primeiras aparições

In [206]:
cubas_novo2 = re.sub('Virgília','Ofélia',cubas,count=2)
cubas_novo2

'Não durou muito a evocação; a realidade dominou logo; o presente expeliu o passado. Talvez eu exponha ao leitor, em algum canto deste livro, a minha teoria das edições humanas. O que por agora importa saber é que Ofélia — chamava-se Ofélia — entrou na alcova, firme, com a gravidade que lhe davam as roupas e os anos, e veio até o meu leito. O estranho levantou-se e saiu. Era um sujeito, que me visitava todos os dias para falar do câmbio, da colonização e da necessidade de desenvolver a viação férrea; nada mais interessante para um moribundo. Saiu; Virgília deixou-se estar de pé; durante algum tempo ficamos a olhar um para o outro, sem articular palavra. Quem diria? De dois grandes namorados, de duas paixões sem freio, nada mais havia ali, vinte anos depois; havia apenas dois corações murchos, devastados pela vida e saciados dela, não sei se em igual dose, mas enfim saciados. Virgília tinha agora a beleza da velhice, um ar austero e maternal; estava menos magra do que quando a vi, pela 

In [207]:
# original
print(len(re.findall('Virgília',cubas)))

# "Virgília" -> "Ofélia" integralmente
print(len(re.findall('Virgília',cubas_novo)))

# "Virgília" -> "Ofélia" em 2 ocorrências apenas
print(len(re.findall('Virgília',cubas_novo2)))

4
0
2


**Exemplo:** Seccione o texto em várias substrings usando a vírgula (",") como ponto de quebra:

In [208]:
cubas_sep = re.split(',',cubas)
cubas_sep

['Não durou muito a evocação; a realidade dominou logo; o presente expeliu o passado. Talvez eu exponha ao leitor',
 ' em algum canto deste livro',
 ' a minha teoria das edições humanas. O que por agora importa saber é que Virgília — chamava-se Virgília — entrou na alcova',
 ' firme',
 ' com a gravidade que lhe davam as roupas e os anos',
 ' e veio até o meu leito. O estranho levantou-se e saiu. Era um sujeito',
 ' que me visitava todos os dias para falar do câmbio',
 ' da colonização e da necessidade de desenvolver a viação férrea; nada mais interessante para um moribundo. Saiu; Virgília deixou-se estar de pé; durante algum tempo ficamos a olhar um para o outro',
 ' sem articular palavra. Quem diria? De dois grandes namorados',
 ' de duas paixões sem freio',
 ' nada mais havia ali',
 ' vinte anos depois; havia apenas dois corações murchos',
 ' devastados pela vida e saciados dela',
 ' não sei se em igual dose',
 ' mas enfim saciados. Virgília tinha agora a beleza da velhice',
 ' um ar

**Exemplo:** Separe a string do texto em no máximo 3 substrings usando o padrão "ão" como ponto de quebra. 

In [209]:
# maxsplit = n; logo, n + 1 substrings
cubas_sep3 = re.split("ão",cubas,maxsplit=2)
cubas_sep3

['N',
 ' durou muito a evocaç',
 '; a realidade dominou logo; o presente expeliu o passado. Talvez eu exponha ao leitor, em algum canto deste livro, a minha teoria das edições humanas. O que por agora importa saber é que Virgília — chamava-se Virgília — entrou na alcova, firme, com a gravidade que lhe davam as roupas e os anos, e veio até o meu leito. O estranho levantou-se e saiu. Era um sujeito, que me visitava todos os dias para falar do câmbio, da colonização e da necessidade de desenvolver a viação férrea; nada mais interessante para um moribundo. Saiu; Virgília deixou-se estar de pé; durante algum tempo ficamos a olhar um para o outro, sem articular palavra. Quem diria? De dois grandes namorados, de duas paixões sem freio, nada mais havia ali, vinte anos depois; havia apenas dois corações murchos, devastados pela vida e saciados dela, não sei se em igual dose, mas enfim saciados. Virgília tinha agora a beleza da velhice, um ar austero e maternal; estava menos magra do que quando 

### Sintaxes especiais

Os metacaracteres podem ser usados para criar REs das formas mais variadas (e criativas) possíveis. Vejamos exemplos.

#### Metacaracter `.`

Serve para combinar qualquer caracter, exceto o _newline_ (`\n`), se usado no modo padrão. Em uso não padrão, pode também combinar-se com `\n`.

**Exemplo:** Localize padrões com 5 caracteres contendo "lh" na 3a. e 4a. posições.

In [210]:
re.findall('..lh.',cubas)

['e lhe', ' olha', 'velhi']

**Exemplo:** Localize padrões iniciando com a letra T maiúscula e contendo, ao todo, 3 caracteres.

In [211]:
re.findall('T...',cubas)

['Talv', 'Tiju']

#### Metacaracter `\`

Serve para criar escapes de caracteres especiais ou sinalizar uma sequência especial.

**Exemplo:** Quebre a string usando o ponto final (".") como ponto de quebra.

In [212]:
# escape do metacaracter '.'
mt1 = re.split('\.',cubas)
mt1

['Não durou muito a evocação; a realidade dominou logo; o presente expeliu o passado',
 ' Talvez eu exponha ao leitor, em algum canto deste livro, a minha teoria das edições humanas',
 ' O que por agora importa saber é que Virgília — chamava-se Virgília — entrou na alcova, firme, com a gravidade que lhe davam as roupas e os anos, e veio até o meu leito',
 ' O estranho levantou-se e saiu',
 ' Era um sujeito, que me visitava todos os dias para falar do câmbio, da colonização e da necessidade de desenvolver a viação férrea; nada mais interessante para um moribundo',
 ' Saiu; Virgília deixou-se estar de pé; durante algum tempo ficamos a olhar um para o outro, sem articular palavra',
 ' Quem diria? De dois grandes namorados, de duas paixões sem freio, nada mais havia ali, vinte anos depois; havia apenas dois corações murchos, devastados pela vida e saciados dela, não sei se em igual dose, mas enfim saciados',
 ' Virgília tinha agora a beleza da velhice, um ar austero e maternal; estava meno

In [213]:
# considera um espaço no padrão
mt2 = re.split('\. ',cubas)
mt2

['Não durou muito a evocação; a realidade dominou logo; o presente expeliu o passado',
 'Talvez eu exponha ao leitor, em algum canto deste livro, a minha teoria das edições humanas',
 'O que por agora importa saber é que Virgília — chamava-se Virgília — entrou na alcova, firme, com a gravidade que lhe davam as roupas e os anos, e veio até o meu leito',
 'O estranho levantou-se e saiu',
 'Era um sujeito, que me visitava todos os dias para falar do câmbio, da colonização e da necessidade de desenvolver a viação férrea; nada mais interessante para um moribundo',
 'Saiu; Virgília deixou-se estar de pé; durante algum tempo ficamos a olhar um para o outro, sem articular palavra',
 'Quem diria? De dois grandes namorados, de duas paixões sem freio, nada mais havia ali, vinte anos depois; havia apenas dois corações murchos, devastados pela vida e saciados dela, não sei se em igual dose, mas enfim saciados',
 'Virgília tinha agora a beleza da velhice, um ar austero e maternal; estava menos magra

#### Metacaracter `^`

Também chamado _caret_, serve para buscar padrões no início de uma string. 

**Exemplo:** Use as substrings do exemplo anterior para buscar quais delas iniciam-se por `'T'`.

In [214]:
for k,s in enumerate(mt2):
    aux = re.match('^T',s)
    if aux: # se lista não for vazia
        print(f'---> Padrão detectado na \
substring {k}:\n\"{s}\"')

---> Padrão detectado na substring 1:
"Talvez eu exponha ao leitor, em algum canto deste livro, a minha teoria das edições humanas"


#### Metacaracter `$`

Visto em sentido oposto a `^`, `$` serve para buscar padrões no final de uma string, ou logo antes de um caracter _newline_.

**Exemplo:** Use as substrings do exemplo anterior para buscar quais delas terminam por `'o'`.

In [215]:
for k,s in enumerate(mt2):
    aux = re.findall('o$',s)
    if aux: # se lista não for vazia
        print(f'---> Padrão detectado na \
substring {k}:\n\"{s}\"')

---> Padrão detectado na substring 0:
"Não durou muito a evocação; a realidade dominou logo; o presente expeliu o passado"
---> Padrão detectado na substring 2:
"O que por agora importa saber é que Virgília — chamava-se Virgília — entrou na alcova, firme, com a gravidade que lhe davam as roupas e os anos, e veio até o meu leito"
---> Padrão detectado na substring 4:
"Era um sujeito, que me visitava todos os dias para falar do câmbio, da colonização e da necessidade de desenvolver a viação férrea; nada mais interessante para um moribundo"


In [216]:
# busca substring terminando com 'undo'
for k,s in enumerate(mt2):
    aux = re.findall('undo$',s)
    if aux: # se lista não for vazia
        print(f'---> Padrão detectado na \
substring {k}:\n\"{s}\"')

---> Padrão detectado na substring 4:
"Era um sujeito, que me visitava todos os dias para falar do câmbio, da colonização e da necessidade de desenvolver a viação férrea; nada mais interessante para um moribundo"


Comentário:

- Note que para buscar o caracter literal '$' em uma string devemos usar o caracter de escape. Vejamos o próximo exemplo.

In [217]:
real = 'João me pagou R$ 150,00.'
v = re.search('\$',real)
print(f'Caracter $ identificado na posição {v.start()}.')

Caracter $ identificado na posição 15.


#### Metacaracter `*`

Serve para buscar 0 ou mais repetições de um caracter precedente. Este caracter especial é de _repetição_.

**Exemplo:**

In [218]:
re.findall('hav*.',cubas)

['ha ', 'ha ', 'ham', 'har', 'havi', 'havi', 'ha ']

Discussão: 

- A RE `'hav*.'` procura por padrões que atendem às seguintes restrições: 
    - o caracter 'v' aparece 0 ou mais vezes após 'ha'.
    - quando não houver 'v', o terceiro caracter passa a ser qualquer um que vier na sequência por causa do metacaracter '.'. 
    - quando houver um 'v', o quarto caracter é qualquer um
    - quando houver n repetições de 'v' após 'a', todos os n são considerados, e o n+1-ésimo caracter é qualquer um.

Comentários: 

- Descrever em palavras o que a expressão regular faz exatamente não é sempre muito fácil. Vejamos mais um exemplo.
    

In [219]:
c = 'ha hahaa hav havv havva-havvva; haiiaa havia'

re.findall('hav*.',c)

['ha ', 'hah', 'hav ', 'havv ', 'havva', 'havvva', 'hai', 'havi']

#### Metacaracter `+`

Serve para buscar 1 ou mais repetições de um caracter precedente. Assim como `*`, `+` também é um caracter especial de _repetição_.

**Exemplo:**

In [220]:
re.findall('hav+.',cubas)

['havi', 'havi']

Discussão: 

- A RE `'hav+.'` é similar à anterior. Porém, pelo fato de ser obrigatória a ocorrência de pelo menos 1 caracter 'v' após 'a', os padrões 'ha ', 'ham' e 'har' são excluídos.
- Se encontradas n ocorrências de 'v', o padrão será encerrado pelo n+1-ésimo caracter arbitrário
- Como no texto existe tal padrão com apenas uma ocorrência de 'v', o quarto caracter, sendo arbitrário, é encontrado como 'i'.
- O padrão 'havia', que forma uma palavra compreensível em Português, é ignorado por conter 5 caracteres.

#### Metacaracter `?`

Serve para buscar 0 ou 1 repetição de um caracter precedente. Assim como `*`, `+` também é um caracter especial de _repetição_.

**Exemplo:**

In [221]:
re.findall('s?an.',cubas)

['ant', 'ana', 'ano', 'anh', 'ant', 'sant', 'ant', 'and', 'ano', 'and']

Comentários: 

- Esta RE procurará por 0 ou 1 repetição de 's' seguida por 'an' e qualquer outro caracter. 

- Neste caso, a substring 'ssan', por exemplo, existente na palavra 'interessante' não seria coberta por `?`.

#### Metacaracteres `[]`

Os colchetes devem ser usados, juntos, em par, para indicar um conjunto de caracteres sobre os quais a expressão regular deve operar. Em geral, usa-se o par `[]` para os seguintes propósitos: 

- Listar caracteres individualmente

A RE `[vnd]` combinará os caracteres `'v'`, `'n'` ou `'d'`.

**Exemplo:**

In [222]:
# localiza todas as ocorrências 
# de 'v', 'n', ou 'd'
print(re.findall('[vnd]',cubas))

['d', 'v', 'd', 'd', 'd', 'n', 'n', 'd', 'v', 'n', 'n', 'd', 'v', 'n', 'd', 'd', 'n', 'v', 'n', 'n', 'v', 'v', 'd', 'd', 'd', 'v', 'n', 'v', 'n', 'v', 'n', 'v', 'v', 'd', 'd', 'd', 'd', 'n', 'd', 'n', 'd', 'd', 'd', 'd', 'n', 'v', 'v', 'v', 'n', 'd', 'n', 'n', 'n', 'd', 'd', 'd', 'd', 'n', 'v', 'd', 'd', 'n', 'd', 'n', 'd', 'd', 'd', 'n', 'd', 'v', 'v', 'n', 'n', 'd', 'v', 'n', 'd', 'd', 'v', 'd', 'v', 'd', 'd', 'd', 'n', 'd', 'n', 'd', 'n', 'd', 'v', 'n', 'v', 'n', 'd', 'n', 'd', 'v', 'v', 'n', 'd', 'n', 'd', 'v', 'n', 'n', 'd']


- Localizar um intervalo de caracteres

Uma RE do tipo `[a-d]` combinará os caracteres de `'a'` a `'d'`, ao passo que `[a-z]` combinará todas as letras minúsculas da tabela ASCII (e não Unicode!), ou seja, de `'a'` a `'z'`.

Do mesmo modo, `[0-4]` combinará os dígitos de 0 a 4, e `[0-9]` todos os dígitos decimais. Por outro lado, a expressão `[0-3][3-8]` combinará todos os números de dois dígitos de 03 a 38.

Se o símbolo `-` for escapado (p.ex. `[a\-d]`), ou se for colocado como primeiro ou último caracter (p.ex. `[-b]` ou `[b-]`), ele significará o hífen literal ('-').

**Exemplo:**

In [223]:
re.findall('[a-z]um',cubas)

['gum', 'hum', 'gum', 'num']

Discussão: 

- Esta RE procurará todos os padrões que iniciam por qualquer letra minúscula (de 'a' a 'z') seguidas por 'um'.

**Exemplo:**

In [224]:
re.findall('[v-z][a-].',cubas)

['va-', 'va,', 'vam', 'van', 'va ', 'zaç', 'vas', 'za ', 'va ', 'vam']

Discussão: 

- Esta RE procurará todos os padrões que iniciam por uma letra minúscula entre 'v' e 'z' seguidas por 'a', '-', ou 'a-' e um caracter qualquer adicional.

**Exemplo:**

In [225]:
# mesmo efeito que anterior
re.findall('[v-z][-a].',cubas)

['va-', 'va,', 'vam', 'van', 'va ', 'zaç', 'vas', 'za ', 'va ', 'vam']

- Localizar caracteres especiais como literais

Os caracteres especiais perdem seu status quando aparecem entre colchetes. Por exemplo, a RE `[(+*)]` procurará por qualquer um dos literais `(`, `+`, `*` ou `)`.

In [226]:
ex = 'Quanto é (22 + 3)*1 - 6*1?'
re.findall('[0-9][)*]',ex)

['3)', '6*']

Discussão:

- Esta RE busca padrões em que um dígito de 0 a 9 precede um ')' ou '*' literalmente.

In [227]:
re.findall('[0-9][?]',ex)

['1?']

Discussão:

- Esta RE busca padrões em que um dígito de 0 a 9 precede um '?' literal.

#### Metacaracteres `{}`

As chaves, assim como os colchetes, devem ser usadas juntas, em par, para localizar cópias ou repetições de uma expressão regular. Em geral, ela é usada para os seguintes objetivos:

- Localizar exatamente _m_ cópias de uma RE anterior

**Exemplo:**

In [228]:
# busca pelo padrão 'rr'
re.search('r{2}',cubas).group()

'rr'

**Exemplo:**

In [229]:
ex2 = '0110011000110011001010110100110111111000'
re.findall('0{2}',ex2)

['00', '00', '00', '00', '00', '00']

In [230]:
re.findall('1{3}',ex2)

['111', '111']

In [231]:
# busca 000000 - 111111
# com variações de 0,1 intermediárias
re.findall('[0-1]{6}',ex2)

['011001', '100011', '001100', '101011', '010011', '011111']

- Localizar entre _m_ e _n_ repetições de uma RE anterior

**Exemplos:**

In [232]:
set(re.findall('10{2,3}',ex2))

{'100', '1000'}

In [233]:
set(re.findall('01{2,3}',ex2))

{'011', '0111'}

In [234]:
set(re.findall('[0-1]{2,3}',ex2))

{'001', '010', '011', '100', '101', '111'}

#### Metacaracter `|`

Usamos o _pipe_ para significar "ou" com o objetivo de localizar mais de uma RE. Por exemplo, `A|B` buscará padrões determinados pela RE `A` **ou** por `B`. _Pipes_ adicionais constroem sintaxes conjuntivas. Ou seja, `A|B|C|D...|Z` significaria "A ou B ou C ou ... D".

**Exemplos:**

In [235]:
# busca por 'rre' ou 'ssa'
re.findall('r{2}e|s{2}a',cubas)

['ssa', 'rre', 'ssa']

In [236]:
# busca por 'cha', 'cho' ou
# 'lha', 'lhi' ou 
# 'pra', 'pre', 'pri', 'pro', 'pru'
re.findall('ch[ao]|lh[ai]|pr[aeiou]',cubas)

['pre', 'cha', 'lha', 'cho', 'lhi', 'pra']

#### Metacaracteres `()`

Servem para criar _grupos de captura_. Um grupo de captura é formado por uma RE confinada entre parênteses. Grupos de captura podem ser aplicados para extrair padrões que possuam uma certa estrutura. 

**Exemplo**:

In [237]:
table = ['alfa00-LAX','beta22-PET', 'zeta92-XIR']

for t in table:
    pat = re.match("([a-z][a-z][a-z][a-z][0-9][0-9])-([A-Z]{3})", t)
    if pat:
        print(pat.groups())

('alfa00', 'LAX')
('beta22', 'PET')
('zeta92', 'XIR')


Discussão: 

- A RE acima é composta de dois grupos de captura que são separados pelo `-` literal. 

- O primeiro grupo de captura é `([a-z][a-z][a-z][a-z][0-9][0-9])`

- O segundo grupo de captura é `([A-Z]{3})`

- Tuplas com dois elementos são impressas com `groups()`.

Comentários: 

- Como se vê, os grupos de captura permitiram que identificássemos padrões separados em uma lista de strings que continha uma estrutura predefinida. 

- Entretanto, as REs anteriores criadas como grupos de captura poderiam ser definidas de uma forma muito mais concisa através de _identificadores_.

#### Identificadores e _tokens_ gerais

Ao trabalhar com REs, podemos utilizar identificadores e _tokens_ diversos para significar padrões de maneira concisa. 

O exemplo anterior sobre grupos de captura, por exemplo, poderia ser escrito de outra forma:

In [238]:
for t in table:
    pat = re.match("(\w+)\W(\w+)", t)
    if pat:
        print(pat.groups())

('alfa00', 'LAX')
('beta22', 'PET')
('zeta92', 'XIR')


Neste caso, `\w` e  `\W` são chamados _identificadores_.

A tabela a seguir resume os principais identificadores e seu significado.

|Identificador|Significado|
|---|---|
|`\d`|qualquer caracter que é um dígito decimal. Equivalente a `[0-9]`|
|`\D`|qualquer caracter exceto dígido (não-dígito). Equivalente a `[^0-9]`|
|`\w`|qualquer caracter Unicode. Se a flag `ASCII` for utilizada, equivale ao conjunto `[a-zA-Z0-9_]`|
|`\W`|o oposto de `\w`. Se a flag `ASCII` for utilizada, equivale ao conjunto `[^a-zA-Z0-9_]`|
|`\s`|qualquer caracter que é um espaço.|
|`\S`|qualquer caracter que não seja espaço.|

A tabela a seguir resume _tokens_ gerais e seu significado.

|_Token_|Significado|
|---|---|
|`\n`|quebra de linha. _Newline_.|
|`\r`|_carriage return_|
|`\t`|TAB|
|`\0`|caracter nulo|

**Exemplo:** Identificando padrões de data e hora. 

In [239]:
dt = ['2021-04-06 10:32:00', '2020-12-03 01:12:58'] 
for m in dt:
    pat = re.match('(\d+)-(\d+)-(\d+)\s(\d+):(\d+):(\d+)',m)
    if pat:
        print(pat.groups())

('2021', '04', '06', '10', '32', '00')
('2020', '12', '03', '01', '12', '58')


**Exemplo:** Identificando cores em padrão hexadecimal.

In [240]:
cores = ['(1,23,43)','#3ed4f4', '#ffcc00','(C,M,Y,K)','#9999ff'] 

for c in cores:
    ok = re.match('#[a-fA-F0-9]{6}',c)
    if ok:
        print(ok.group())

#3ed4f4
#ffcc00
#9999ff


**Exemplo:** Identificando URLs no domínio Wikipedia Português.

In [241]:
urls = ['http://pt.wikipedia.org/wiki/Bras',
       'http://wikipedia.org/wiki/Bras',
       'https://pt.wikipedia.org/wiki/Bras',
       'http://pt.wikipedia.org/wik/Bras',
       'https://pt.wikipedia.org/wiki/Cubas']

for u in urls:
    ok = re.match('((http|https)://)?(pt.wikipedia.org/wiki/).+',u)
    if ok:
        print(ok.group())

http://pt.wikipedia.org/wiki/Bras
https://pt.wikipedia.org/wiki/Bras
https://pt.wikipedia.org/wiki/Cubas


#### Compilação e literais

Até agora abordamos REs de uma maneira direta, mostrando como podemos localizar padrões pela utilização de alguns métodos. Vale comentar que REs podem ser compiladas como objetos mais abstratos. 

Para compilar expressões, podemos utilizar o método `compile`, no qual REs são manipuladas como strings. 

**Exemplo:**

In [242]:
rex = re.compile('[p-t]a[b-d]a?')
rex.findall(cubas)

['sad', 'sab', 'rad', 'tad', 'sac', 'sac']

Aqui, `rex` é um objeto abstrato.

In [243]:
print(type(rex))

<class 're.Pattern'>


Se imprimirmos `rex`, veremos o seguinte:

In [244]:
rex

re.compile(r'[p-t]a[b-d]a?', re.UNICODE)

O prefixo `r` torna a string literal e serve para tratar problemas com escape. Todavia, deixaremos esta discussão para um momento posterior. Para saber mais, leia sobre a [Praga do Backspace](https://docs.python.org/3/howto/regex.html#the-backslash-plague).

#### Recursos de aprendizagem

Para aprofundar seu entendimento sobre expressões regulares, recomendamos os seguintes sites: 

- [regex101.com](https://regex101.com)
- [regexr.com](https://regexr.com)
- [Regular Expression HOWTO](https://docs.python.org/3/howto/regex.html)
- [Documentação do módulo `re`](https://docs.python.org/3/library/re.html)